Source/ResponseSerialization.swift

```diff
@@ -303,19 +308,40 @@ extension DataRequest {
 /// request returning `nil` or no data is considered an error. However, if the response is has a status code valid for
 /// empty responses (`204`, `205`), then an empty `Data` value is returned.
 public final class DataResponseSerializer: ResponseSerializer {
+    /// HTTP response codes for which empty responses are allowed.
+    public let emptyResponseCodes: Set<Int>
+    /// HTTP request methods for which empty responses are allowed.
+    public let emptyRequestMethods: Set<HTTPMethod>
 
-    public init() { }
+    /// Creates an instance using the provided values.
+    ///
+    /// - Parameters:
+    ///   - emptyResponseCodes:  The HTTP response codes for which empty responses are allowed. Defaults to
+    ///                          `[204, 205]`.
+    ///   - emptyRequestMethods: The HTTP request methods for which empty responses are allowed. Defaults to `[.head]`.
+    public init(emptyResponseCodes: Set<Int> = DataResponseSerializer.defaultEmptyResponseCodes,
+                emptyRequestMethods: Set<HTTPMethod> = DataResponseSerializer.defaultEmptyRequestMethods) {
+        self.emptyResponseCodes = emptyResponseCodes
+        self.emptyRequestMethods = emptyRequestMethods
+    }
 
     public func serialize(request: URLRequest?, response: HTTPURLResponse?, data: Data?, error: Error?) throws -> Data {
         guard error == nil else { throw error! }
-
-        if let response = response, emptyDataStatusCodes.contains(response.statusCode) { return Data() }
-
-        guard let validData = data else {
-            throw AFError.responseSerializationFailed(reason: .inputDataNil)
+        
+        guard let data = data, !data.isEmpty else {
+            let requestAllowsEmptyResponses = request.flatMap { $0.httpMethod }
+                                                     .flatMap(HTTPMethod.init)
+                                                     .map { emptyRequestMethods.contains($0) }
+            let responseAllowsEmptyResponses = response.flatMap { $0.statusCode }
+                                                       .map { emptyResponseCodes.contains($0) }
+            if requestAllowsEmptyResponses ?? responseAllowsEmptyResponses ?? false {
```

Comment:
This is a great change overall! I saw some potential for reuse here given that each of the four response serializers repeated this logic. I took a stab at moving it all into the `ResponseSerializer` protocol and the protocol extension in PR #2646. It makes this logic more reusable for custom response serializers both inside and outside AF. Let me know what you think.

Reply:
Updated.